### PR TITLE
Merge development into main

### PR DESCRIPTION
**Pull request by jlkravitz**
- [x] stand up the site locally
	- [x] test all functionality in all major browsers, emphasizing the functionality that this pull request addresses
		- [x] for public-facing functionality, test in browsers consistent with [public browser use data](https://analytics.usa.gov/)
		- [x] test in Mobile Safari and Mobile Chrome
	- [x] use an automated audit tool for code quality and practices (recommended: [Chrome DevTools](https://developers.google.com/web/tools/chrome-devtools/), aka [Lighthouse](https://chrome.google.com/webstore/detail/lighthouse/blipmdconlkpinefehnmjammfjpmpbjk))
		- [x] look at efficiency of page loads, asset sizes, HTTP connection management, etc.
	- [x] review for accessibility
		- [x] use an automated audit tool, such as Chrome Audit or aXe
		- [x] navigate site only with the keyboard
		- [x] use VoiceOver or Narrator to navigate the site with audio only, with the display turned off
		- [x] manually test anything that pa11y cannot test automatically (e.g., contrast of text over images)
- [x] review static code analysis results, if available
- [x] run a security audit of dependencies (e.g. `npm audit` and `pip audit`) to ensure that there are no vulnerabilities that will be deployed to production (as opposed to vulnerabilities that only have an impact on the development environment)
- [x] examine OWASP ZAP output to ensure that any errors are known to be false positives or have been previously declared to be acceptable
- [x] for each feature-level bug (i.e., it’s working as designed, but designed wrong), open a new issue and put it in the backlog

## Potential Issues
- [x] The frontend does not display an error when the user is unauthorized to upload to a particular jurisdiction. (Instead, nothing happens.) **Should be fixed by #184.**
- [x] The frontend does not restrict the bulk upload jurisdiction list to those jurisdictions to which I have write permissions. **See #147**
- [x] I cannot navigate the provider list with my keyboard.
- [x] How do I set up authentication / Cognito with a sandbox deployment? I'm getting errors when I use the test app callback URL.
- [x] How would you recommend running pa11y or ZAP automated testing now that the main pages are authenticated?
- [x] When I test on additional browsers, Cognito gives me an authentication error.
- [x] While logged in, when I open a page in a new tab, I am logged out and presented with the Cognito log-in screen.
- [x] Pa11y: #142

### CODE REVIEW

**Comment by isabeleliassen** [comment]
Hey @jlkravitz Is it a problem that some of the checks for this failed?

**Comment by jusdino** [comment]
Interesting. That looks like a new false-positive from `pylint`. I can add a suppression for it.

File: backend/compact-connect/docs/README.md
Comment (jlkravitz) [inline backend/compact-connect/docs/README.md:35]: # General Comments & Questions

* The frontend does not display an error when the user is unauthorized to upload to a particular jurisdiction. @jsandoval81 
* The frontend does not restrict the bulk upload jurisdiction list to those jurisdictions to which I have write permissions. @jsandoval81 
* How do I set up authentication / Cognito with a sandbox deployment? I'm getting errors when I use the test app callback URL. @jusdino 
* How would you recommend running pa11y or ZAP automated testing now that the main pages are authenticated? @jusdino

**Comment by jsandoval81** [comment]
@jlkravitz 

- > _The frontend does not display an error when the user is unauthorized to upload to a particular jurisdiction. (Instead, nothing happens.)_
    - I think the latest backend PR merge actually had some breaking updates, so v0 may not be stable. If you're still seeing this in the [open frontend PR](https://github.com/csg-org/CompactConnect/pull/184) I can address it there. Though for me, it's showing an "Access Denied" message:
    - Screenshot 2024-09-03 at 9 55 02 AM
- > _The frontend does not restrict the bulk upload jurisdiction list to those jurisdictions to which I have write permissions._
    - We won't have that ability until after we get the user management endpoints, which is currently in-progress (#147)
    - Does remind me that we'll also need an update to Figma to call that UI feature out 👍

File: backend/compact-connect/docs/README.md
Comment (jusdino) [reply backend/compact-connect/docs/README.md:35]: > How do I set up authentication / Cognito with a sandbox deployment? I'm getting errors when I use the test app callback URL. @jusdino

There is a script under `backend/compact-connect/bin` to create yourself a user. After you've run that, you should be able to log in. The next thing would be making sure that the Cognito app client is configured to allow the callback to your UI. That part depends on where you're running the UI. If you're running the UI from CloudFront with a domain name, it should work. If you're running the UI locally, you'll need to make sure that you have the context variable `"allow_local_ui": true` set for your sandbox environment (included in `cdk.context.sandbox-example.json`. That flag will allow authorization callbacks to `"http://localhost:3018/auth/callback"`.

> How would you recommend running pa11y or ZAP automated testing now that the main pages are authenticated? @jusdino

Each of those tools supports authentication in their own way - here are a couple of references that offer a hint:
- https://www.zaproxy.org/docs/desktop/start/features/authentication/
- https://github.com/pa11y/pa11y/issues/242

If you think it's time for us to automate those scans, we should probably talk to @isabeleliassen about getting some tickets in the backlog. We can also talk about disabling MFA requirements in `test` (or your sandbox), if MFA is getting in the way of efficient testing for you.

> When I test on additional browsers, Cognito gives me an authentication error. @jusdino

It looks like you're running afoul of Cognito's intrusion detection / advanced security monitoring. If Cognito keeps shutting you out, like MFA, we can talk about loosening up Cognito's security configuration for a pre-production environment.

Comment (jlkravitz) [reply backend/compact-connect/docs/README.md:35]: > > How do I set up authentication / Cognito with a sandbox deployment? I'm getting errors when I use the test app callback URL. @jusdino
> 
> There is a script under `backend/compact-connect/bin` to create yourself a user. After you've run that, you should be able to log in. The next thing would be making sure that the Cognito app client is configured to allow the callback to your UI. That part depends on where you're running the UI. If you're running the UI from CloudFront with a domain name, it should work. If you're running the UI locally, you'll need to make sure that you have the context variable `"allow_local_ui": true` set for your sandbox environment (included in `cdk.context.sandbox-example.json`). That flag will allow authorization callbacks to `"http://localhost:3018/auth/callback"`.
> 
Helpful, thank you! I'll try this and return with any follow-ups.

> > How would you recommend running pa11y or ZAP automated testing now that the main pages are authenticated? @jusdino
> 
> Each of those tools supports authentication in their own way - here are a couple of references that offer a hint:
> 
> * https://www.zaproxy.org/docs/desktop/start/features/authentication/
> * [pa11y authentication and login pa11y/pa11y#242](https://github.com/pa11y/pa11y/issues/242)
> 
> If you think it's time for us to automate those scans, we should probably talk to @isabeleliassen about getting some tickets in the backlog. We can also talk about disabling MFA requirements in `test` (or your sandbox), if MFA is getting in the way of efficient testing for you.
> 

Right, I read through a lot of this but was having trouble setting up and didn't want to spend too much time on it. I think partly it can't easily deal with MFA, so perhaps disabling MFA in `test` or my sandbox is a good first step. North star is still to automate these scans, of course, but if we can't get them running manually, I doubt we'll be able to get them running automatically anyways.

> > When I test on additional browsers, Cognito gives me an authentication error. @jusdino
> 
> It looks like you're running afoul of Cognito's intrusion detection / advanced security monitoring. If Cognito keeps shutting you out, like MFA, we can talk about loosening up Cognito's security configuration for a pre-production environment.

Got it. Definitely a blocker as I'm unable to test on various browsers. I'm inclined to prefer identical setups on prod and pre-prod, but if this is the only reasonable option, then so be it. FWIW, I didn't even get to the MFA screen when I received these errors, so it feels like something that's happening _before_ MFA.

**Comment by jlkravitz** [comment]
>   * I think the latest backend PR merge actually had some breaking updates, so v0 may not be stable. If you're still seeing this in the [open frontend PR](https://github.com/csg-org/CompactConnect/pull/184) I can address it there. Though for me, it's showing an "Access Denied" message:
>   * Screenshot 2024-09-03 at 9 55 02 AM

Got it!

> * > _The frontend does not restrict the bulk upload jurisdiction list to those jurisdictions to which I have write permissions._
>   
>   
>   
>   * We won't have that ability until after we get the user management endpoints, which is currently in-progress ([backend API for CSG staff user management #147](https://github.com/csg-org/CompactConnect/issues/147))
>   * Does remind me that we'll also need an update to Figma to call that UI feature out 👍

Makes sense. Just something I noticed.

> I cannot navigate the provider list with my keyboard. @jsandoval81

And any insight on this one?

**Comment by jlkravitz** [comment]
@jusdino @jsandoval81 

Another note in my testing. When I open the upload or license page in a new tab, it logs me out and requires that I log in again.

I know there's a lot of back-and-forth here, so I'm going to update the PR description with the issues I brought up.

**Comment by jsandoval81** [comment]
@jlkravitz 

- > _While logged in, when I open a page in a new tab, I am logged out and presented with the Cognito log-in screen._
    - This is expected. We're currently using `sessionStorage` for login persistence - which is isolated per tab. It's simply the more conservative approach since `sessionStorage` is cleared when the tab is closed. If the expectation is that login should persist across multiple tabs in a single browser, then I'll just need the OK from you to use `localStorage` (which is never automatically cleared) and I can create a backlog item for that.
- > _I cannot navigate the provider list with my keyboard_
    - Good catch. I can update that in the current frontend PR.
- https://github.com/csg-org/CompactConnect/issues/142
    - I can probably also fix this in the current frontend PR.

**Comment by jlkravitz** [comment]
> * This is expected. We're currently using `sessionStorage` for login persistence - which is isolated per tab. It's simply the more conservative approach since `sessionStorage` is cleared when the tab is closed. If the expectation is that login should persist across multiple tabs in a single browser, then I'll just need the OK from you to use `localStorage` (which is never automatically cleared) and I can create a backlog item for that.
> * >

Ah, interesting. Let's table for now. It's certainly not the UX I would expect, but let's discuss later the security pros/cons.

**Comment by jlkravitz** [comment]
@jsandoval81 Two questions before we merge this–

1. It seems the filters only show up on mobile. Is that intended?
2. When they do appear, I can't seem to open and set the filters using my keyboard. 

Desktop: 
image

Mobile:
image

**Comment by jsandoval81** [comment]
@jlkravitz 

1. That's intended. That's just the alternative UI for sorting on mobile; since the common column headers aren't visible on mobile.
2. I created a draft item in the backlog to add keyboard nav to that mobile-only element. Hopefully it's ok to address that in an upcoming sprint and not hold up this merge / deploy.

Let me know if you have any follow-up thoughts on that.

**Comment by jlkravitz** [comment]
> @jlkravitz
> 
> 1. That's intended. That's just the alternative UI for sorting on mobile; since the common column headers aren't visible on mobile.
> 2. I created a draft item in the backlog to add keyboard nav to that mobile-only element. Hopefully it's ok to address that in an upcoming sprint and not hold up this merge / deploy.
> 
> Let me know if you have any follow-up thoughts on that.

Yup both good with me. Missed the column sorting capability!

@isabeleliassen good to merge this.

**Review by isabeleliassen — APPROVED** [review]

**Pull request merged**
